File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:

        self.ans = 0

        def dfs(node):
            if node is None:
                return (0, 0)   # (sum, count)

            # Get left subtree sum and count
            leftSum, leftCount = dfs(node.left)

            # Get right subtree sum and count
            rightSum, rightCount = dfs(node.right)

            # Current subtree sum and count
            totalSum = node.val + leftSum + rightSum
            totalCount = 1 + leftCount + rightCount

            # Calculate average
            average = totalSum // totalCount

            # Check if node value equals average
            if node.val == average:
                self.ans += 1

            return (totalSum, totalCount)

        dfs(root)

        return self.ans
        

File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;

    pair<int, int> dfs(TreeNode* root) {
        if (root == nullptr) {
            return {0, 0};   // {sum, count}
        }

        // Get left subtree sum and count
        auto left = dfs(root->left);

        // Get right subtree sum and count
        auto right = dfs(root->right);

        // Calculate current subtree sum and count
        int sum = root->val + left.first + right.first;
        int count = 1 + left.second + right.second;

        // Calculate average
        int average = sum / count;

        // Check if current node equals average
        if (root->val == average) {
            ans++;
        }

        return {sum, count};
    }

    int averageOfSubtree(TreeNode* root) {
        dfs(root);
        return ans;
    }
};